### PR TITLE
Add a TransportExt trait

### PR DESCRIPTION
**Pull request by tomaka**

### CODE REVIEW

**Comment by tomaka** [comment]
Tests are failing because of a compiler ICE. I can't reproduce it locally, so I assume it's been fixed in the nightly version of rustc.

**Comment by tomaka** [comment]
I commented out the `with_rate_limit` function which causes the ICE.

File: src/transport_ext.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    fn with_timeout(self, timeout: Duration) -> TransportTimeout<Self>
+    where
+        Self: Sized,
+    {
+        TransportTimeout::new(self, timeout)
+    }
+
+    /// Adds a timeout to all the outgoing sockets created by the transport.
+    #[inline]
+    fn with_outgoing_timeout(self, timeout: Duration) -> TransportTimeout<Self>
```
Comment (dvdplm) [inline src/transport_ext.rs]: I think `outgoing` is an unfortunate choice of words here. How about `with_outbound_timeout`? "outgoing" also means "extrovert" which reads a bit funny.

(Same for `ingoing` below, not really a word – how about "inbound"?).

File: src/transport_ext.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    fn with_ingoing_timeout(self, timeout: Duration) -> TransportTimeout<Self>
+    where
+        Self: Sized,
+    {
+        TransportTimeout::with_ingoing_timeout(self, timeout)
+    }
+
+    // TODO: this method causes an ICE in Rust 1.29 but works in Rust 1.30 ; restore it when
+    //       1.30 is released
+    /*/// Adds a maximum transfert rate to the sockets created with the transport.
```
Comment (dvdplm) [inline src/transport_ext.rs]: s/transfert/transfer/

Does the rate limit apply to both in- and outbound sockets? I'd assume it's only for inbound traffic, but might be good to specify, so maybe: "Sets a maximum transfer rate on inbound sockets created with the transport."

File: src/transport_ext.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    #[inline]
+    fn with_rate_limit(
+        self,
+        max_read_per_sec: usize,
+        max_write_per_sec: usize,
```
Comment (dvdplm) [inline src/transport_ext.rs]: Are the `max_read` and `max_write` limits referring to the number of payloads (i.e. number of reads/writes of any size) or the number of bytes read/written?

File: src/transport_ext.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+///
+/// # Example
+///
+/// ```
+/// use libp2p::TransportExt;
+/// use libp2p::tcp::TcpConfig;
+/// use std::time::Duration;
+///
+/// let _transport = TcpConfig::new()
+///     .with_timeout(Duration::from_secs(20));
```
Comment (dvdplm) [inline src/transport_ext.rs]: When does the clock start ticking, on first poll? On first read/write?

File: src/transport_ext.rs (hunk 2 above)
Comment (tomaka) [reply src/transport_ext.rs]: > I'd assume it's only for inbound traffic, but might be good to specify

Why do you assume it's only for inbound traffic?

File: src/transport_ext.rs (hunk 3 above)
Comment (tomaka) [reply src/transport_ext.rs]: The size of a payload is totally unpredictable, so I don't see how using a number of payloads would be a good idea.

File: src/transport_ext.rs (hunk 4 above)
Comment (tomaka) [reply src/transport_ext.rs]: It's a timeout for when you connect, not when you read/write.

**Comment by tomaka** [comment]
Restored the `with_rate_limit` method.
This will cause the build to fail until Rust 1.30 is released.

File: src/transport_ext.rs (hunk 2 above)
Comment (dvdplm) [reply src/transport_ext.rs]: Well, it's common to see inbound traffic as a possible cause for trouble and worth controlling/limiting while leaving outbound traffic free. If that is not the case then maybe that's worth mentioning explicitly in the docs?

Comment (tomaka) [reply src/transport_ext.rs]: > If that is not the case then maybe that's worth mentioning explicitly in the docs?

That's what I did in the latest commit.

File: src/transport_ext.rs (hunk 3 above)
Comment (dvdplm) [reply src/transport_ext.rs]: As an example of "per payload" rate limiting, Nginx rate limiting works on the number of requests (of any kind/size). Anyway: these are bytes then.

**Pull request merged**